File: Java/Chapter1/1_2_Check_Permutation.java
//O(N)

public class Main {
  public static boolean perm(String str1, String str2) {
    if (str1.length() != str2.length()) {
      return false;
    }
    
    int[] letters = new int[128];
    
    for(int i = 0; i < str1.length(); i++) {
      letters[str1.charAt(i)] += 1;
    }
    
    for(int i = 0; i < str2.length(); i++) {
      if (letters[str2.charAt(i)] < 0) {
        return false;
      }
      letters[str2.charAt(i)] -= 1;
    }
    
    return true;
  }
  
  public static void main(String[] args) {
    String str1 = "abhinav";
    String str2 = "abhivan";
    System.out.println(perm(str1, str2));
  }
}

File: Python/Chapter1/1_2_Check_Permutation.py
# O(N)

def perm(str1, str2):
  if len(str1) != len(str2):
    return False
  
  d = {}
  
  for char in str1:
    if char in d:
      d[char] += 1
      
    else:
      d[char] = 1
      
  for char in str2:
    if d[char] == 0:
      return False
    else:
      d[char] -= 1
  
  return True
      
      
print(perm('abhinav', 'abhivan'))
